stable_baselines/trpo_mpi/trpo_mpi.py

```diff
@@ -444,19 +444,36 @@ def predict(self, observation, state=None, mask=None, deterministic=False):
             state = self.initial_state
         if mask is None:
             mask = [False for _ in range(self.n_envs)]
-        observation = np.array(observation).reshape((-1,) + self.observation_space.shape)
+        observation = np.array(observation)
+        vectorized_env = self._is_vectorized_observation(observation, self.observation_space)
 
+        observation = observation.reshape((-1,) + self.observation_space.shape)
         actions, _, states, _ = self.step(observation, state, mask, deterministic=deterministic)
+
+        if not vectorized_env:
+            if state is not None:
+                raise ValueError("Error: The environment must be vectorized when using recurrent policies.")
+            actions = actions[0]
+
         return actions, states
 
     def action_probability(self, observation, state=None, mask=None):
         if state is None:
             state = self.initial_state
         if mask is None:
             mask = [False for _ in range(self.n_envs)]
-        observation = np.array(observation).reshape((-1,) + self.observation_space.shape)
+        observation = np.array(observation)
+        vectorized_env = self._is_vectorized_observation(observation, self.observation_space)
```

Comment:
Looks like duplicated code (see A2C/ACER, ...) I think we could make a common method for that

Reply:
I think we should just split the BaseRLModel into ActorCriticRLModel and OffPolicyRLModel (DQN & DDPG). This should reduce a lot of redudent code (not just the predict functions), not to mention this will make integrating HER a lot easier.

Reply:
I agree, maybe more OnePolicy vs OffPolicy to be consistent, no ?